framework/src/tinyalsa/tinyalsa.c

```diff
@@ -614,7 +625,7 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 				if (ioctl(pcm->fd, AUDIOIOC_ENQUEUEBUFFER, (unsigned long)&bufdesc) < 0) {
 					return oops(pcm, errno, "failed to enque buffer after read");
 				}
-				pcm->buf_ptr++;
+				apb->flags |= AUDIO_APB_OUTPUT_ENQUEUED;
```

Comment:
Should we do this in tinyalsa layer, not audio layer?????

Reply:
I think its better to do in tinyalsa. 
If we do in audio layer and there is some problem in the msg queue, then it wont be valid.
If we do at this point in tinyalsa, we are sure that enqueue/dequeue has completed successfully.

Reply:
In my opinion, handling buffer(change something in buffer structure) should be done by audio layer.
I think we can change it in device layer before send ret to upper layer.
Please double check this.

Reply:
Yes. I understand your point about setting the flags in audio layer. 
But, I am using the AUDIO_APB_OUTPUT_ENQUEUED flag to check if the buffer is available to tinyalsa for writing data. Suppose, the audio layer sets AUDIO_APB_OUTPUT_ENQUEUED. But the buffer is still in queue. Then, it will lead to mistakes. So, I want to update the flag only after removing the buffer from the queue.

Reply:
What I mean is that change state right before ioctl return & mqueue returns dequeue result.

Reply:
You mean that we have to change the state in audio.c, before return from ioctl. But, when we return from ioctl, the buffer is still in the queue. I want to change the state after removing the buffer from the audio msg queue. Only then, the buffer is available to write data.

Reply:
I don't understand well, in your code, tinyalsa request enqueue through the ioctl, and ioctl request enqueueing buffer to i2s, whenever callback function called from i2s, it changed apb state.
Please check alc5658_enqueuebuffer that even if priv is not running, it returns ok..
In other words, AUDIO_APB_OUTPUT_ENQUEUED should be set after driver request i2s_send or i2s_receiver, on the other hands, AUDIO_APB_OUTPUT_ENQUEUED should be unset driver's call back called with valid buffer data.
Make sure that only driver knows it enqued or dequeued properly...

Reply:
My previous comment is a bit confusing / misleading!! Sorry.
What I mean is this. I am using the AUDIO_APB_OUTPUT_ENQUEUED flag to check whether the buffer is with tinyalsa (means we can use it to read / write) OR the buffer is not with tinyalsa (its with codec or i2s or mqueue) and so we cannot use it. I dont think there is any need for driver to use this flag to track the buffer state. It will be used only in tinyalsa to check if the buffer is in tinyalsa or not. If its not in tinyalsa, then it does not matter if it is in codec or i2s.
That is why I set the flag on a successful return from ioctl(AUDIOIOC_ENQUEUEBUFFER) and reset it on successfully receiving AUDIO_MSG_DEQUEUE. This flag is not being used or checked anywhere else in the code.

Reply:
What I am saying is that ioctl return ok without enqueing, in this case tinyalsa set buffer state as ENQUEUED.
That's why I ask to you double check.
If it used in tinyalsa to check buffer is enqueued or something, we should make an extra policy.
(For example, ioctl in driver returns size of buffer if enqueued. or define new constatnt (for example, AUDIOIOC_GET_BUFFERSTATE)to use extra ioctl to get buffer state...